code/game/objects/structures/mortar.dm

```diff
@@ -9,109 +9,110 @@
-			return
-		if(busy)
-			to_chat(user, "<span class='warning'>Someone else is currently using this mortar.</span>")
+	if(user.skills.getRating("engineer") < SKILL_ENGINEER_ENGI)
+		user.visible_message("<span class='notice'>[user] fumbles around figuring out how to use [src].</span>",
+		"<span class='notice'>You fumble around figuring out how to use [src].</span>")
+		var/fumbling_time = 4 SECONDS * ( SKILL_ENGINEER_ENGI - user.skills.getRating("engineer") )
+		if(!do_after(user, fumbling_time, TRUE, src, BUSY_ICON_UNSKILLED))
 			return
-		user.visible_message("<span class='notice'>[user] starts dialing [src]'s firing angle and distance.</span>",
-		"<span class='notice'>You start dialing [src]'s firing angle and distance to match the new coordinates.</span>")
-		busy = 1
-		playsound(loc, 'sound/items/ratchet.ogg', 25, 1)
-		if(do_after(user, 15, TRUE, src, BUSY_ICON_GENERIC))
-			user.visible_message("<span class='notice'>[user] finishes dialing [src]'s firing angle and distance.</span>",
-			"<span class='notice'>You finish dialing [src]'s firing angle and distance to match the new coordinates.</span>")
-			busy = 0
-			dial_x = temp_dial_x
-			dial_y = temp_dial_y
-		else busy = 0
+	ui_interact(user)
+
 
+/obj/structure/mortar/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "Mortar", "M402 Mortar", 430, 150, master_ui, state)
+		ui.open()
+
+/obj/structure/mortar/ui_data(mob/user)
+	. = ..()
+	var/data = list()
+	data["X"] = coords["targ_x"]
+	data["Y"] = coords["targ_y"]
+	data["D1"] = coords["dial_x"]
+	data["D2"] = coords["dial_y"]
+	data["last_three_inputs"] = last_three_inputs
+	return data
+
+/obj/structure/mortar/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
+	. = TRUE
+
+	switch(action)
+		if("change_target_x")
+			coords["targ_x"] = clamp(text2num(params["target_x"]), 0, world.maxx)
+		if("change_target_y")
+			coords["targ_y"] = clamp(text2num(params["target_y"]), 0, world.maxy)
+		if("change_dial_x")
+			coords["dial_x"] = clamp(text2num(params["dial_one"]), -10, 10)
+		if("change_dial_y")
+			coords["dial_y"] = clamp(text2num(params["dial_two"]), -10, 10)
+		if("set_saved_coord_one")
+			coords = get_new_list("coords_one")
+		if("set_saved_coord_two")
+			coords = get_new_list("coords_two")
+		if("set_saved_coord_three")
+			coords = get_new_list("coords_three")
+	if((coords["targ_x"] != 0 && coords["targ_y"] != 0))
+		usr.visible_message("<span class='notice'>[usr] adjusts [src]'s firing angle and distance.</span>",
+		"<span class='notice'>You adjust [src]'s firing angle and distance to match the new coordinates.</span>")
+		playsound(loc, 'sound/items/ratchet.ogg', 25, 1)
+/obj/structure/mortar/proc/get_new_list(str)
+	var/list/target_data = list()
```

Comment:
```suggestion
	if((coords["targ_x"] != 0 && coords["targ_y"] != 0))
		usr.visible_message("<span class='notice'>[usr] adjusts [src]'s firing angle and distance.</span>",
		"<span class='notice'>You adjust [src]'s firing angle and distance to match the new coordinates.</span>")
		playsound(loc, 'sound/items/ratchet.ogg', 25, 1)

/obj/structure/mortar/proc/get_new_list(str)
	var/list/target_data = list()
```

Reply:
yeah no I want it to be on the exact moment they enter in the values, not when you're firing

Reply:
What are you talking about this only adds one tab

Reply:
you mean removes?

Reply:
![image](https://user-images.githubusercontent.com/57223640/95611536-41168600-0a62-11eb-91ad-47846adfff7b.png)